### PR TITLE
[WIP] Bump boost to v1.70.0

### PR DESCRIPTION
**Pull request by dberzano**

### CODE REVIEW

**Comment by pzhristov** [comment]
Hi @dberzano,

Does DDS compile with Boost 1.69? I had some issues since they requested the deprecated signals library from Boost.

Cheers,
Peter

**Comment by dberzano** [comment]
No, it does not, for exactly the same reason. I have opened the PR as a reference, I believe we need to solve this problem together with the DDS developers.

**Comment by awegrzyn** [comment]
@dberzano Brew does not ship v1.68 anymore (just v1.69 or 1.60), what's the recommended of getting boost for a new macOS developer?

**Comment by pzhristov** [comment]
Hi @dberzano and @awegrzyn,
I was able to compile DDS with Boost 1.69 on Mac with Xcode 10.2, but this version has internal problems. Two header files have to be modified (I edited them directly on my Mac, but the real patch should be system dependent. We have two problems in the files wait_for_exit.hpp and wait_group.hpp:

- sigemptyset(...) and sigaddset(...) are used from the global namespace, but on Mac they are not there, so I removed the :: in front of the calls;

- there is a call to sigtimedwait(...), but this function is not implemented on Mac, so I commented out the part where the call is done and the result is used.

If we have Boost 1.69 in our repository, I can try to provide a Mac-specific patch.

Cheers,
Peter

**Comment by dberzano** [comment]
@pzhristov please create a PR to this branch (not master): https://github.com/alisw/boost/tree/alice/v1.69.0-patches

**Comment by Barthelemy** [comment]
l confirm that DDS does not compile with boost 1.69. 
This is becoming a serious issue as brew updates to 1.69. 
@AnarManafov @AndreyLebedev could you have a look ?

**Comment by AnarManafov** [comment]
@Barthelemy , we have fixed the boost_signals issue 3 months ago (https://github.com/FairRootGroup/DDS/commit/52dd0097c98a2f30030b749edfd21a051f0cd38e)

**Comment by ktf** [comment]
@AnarManafov do you have a release for it?

**Comment by Barthelemy** [comment]
@AnarManafov Thanks, I did not know.

File: boost.sh [hunk 1]
```diff
@@ -11,7 +11,7 @@ build_requires:
   - "bz2"
 prefer_system: (?!slc5)
 prefer_system_check: |
-  printf "#include \"boost/version.hpp\"\n# if (BOOST_VERSION < 106800 || BOOST_VERSION > 106899)\n#error \"Cannot use system's boost: boost 1.68 required.\"\n#endif\nint main(){}" | c++ -I$(brew --prefix boost)/include -xc++ - -o /dev/null
+  printf "#include \"boost/version.hpp\"\n# if (BOOST_VERSION < 106900 || BOOST_VERSION > 106999)\n#error \"Cannot use system's boost: boost 1.69 required.\"\n#endif\nint main(){}" | gcc -I$(brew --prefix boost)/include -xc++ - -o /dev/null
```
Comment (ktf) [inline boost.sh]: Do not change this to gcc.

**Comment by AnarManafov** [comment]
> @AnarManafov do you have a release for it?

@ktf , no, not yet. It's in the master.
We wanted to add more features, before we tag a new release.
But the current master is stable. If it's absolutly critical for you, we can make a stable release.

**Comment by ktf** [comment]
Is that commit the only one needed or there is more? I think we can live with a prerelease of 2.4, assuming everything builds.

**Comment by AnarManafov** [comment]
BTW, I am not sure it is still true, but I thin FairMQ also had issue building against boost 1.69.
This is why we decided to skip this version.

**Comment by AnarManafov** [comment]
> Is that commit the only one needed or there is more? I think we can live with a prerelease of 2.4, assuming everything builds.

This is the only change needed. We don't actually use boot signals v1 in the code. We was just accidentally linking against it. 
In the code we use boost signals v2 and it is a header only lib.

**Comment by MohammadAlTurany** [comment]
> BTW, I am not sure it is still true, but I thin FairMQ also had issue building against boost 1.69.
> This is why we decided to skip this version.

There was issues with MacOS and Boost 1.69 and not with FairMQ

**Comment by pzhristov** [comment]
> > BTW, I am not sure it is still true, but I thin FairMQ also had issue building against boost 1.69.
> > This is why we decided to skip this version.
> 
> There was issues with MacOS and Boost 1.69 and not with FairMQ

Yes, exactly. I am testing a patch for the Boost 1.69 issue on Mac.

**Comment by awegrzyn** [comment]
@AnarManafov DDS master fails with boost 1.69 provided by brew with:
```
Scanning dependencies of target dds_user_defaults_lib
[  1%] Building CXX object dds-user-defaults/CMakeFiles/dds_user_defaults_lib.dir/src/UserDefaults.cpp.o
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/dds-user-defaults/src/UserDefaults.cpp:23:
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/MiscCommon/FindCfgFile.h:11:
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/MiscCommon/SysHelper.h:33:
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/MiscCommon/MiscUtils.h:14:
In file included from /usr/local/opt/boost/include/boost/process.hpp:24:
In file included from /usr/local/opt/boost/include/boost/process/async_system.hpp:22:
In file included from /usr/local/opt/boost/include/boost/process/child.hpp:21:
In file included from /usr/local/opt/boost/include/boost/process/detail/child_decl.hpp:30:
/usr/local/opt/boost/include/boost/process/detail/posix/wait_for_exit.hpp:60:7: error: expected unqualified-id
    ::sigemptyset(&sigset);
      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/signal.h:125:26: note: expanded from macro 'sigemptyset'
#define sigemptyset(set)        (*(set) = 0, 0)
                                ^
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/dds-user-defaults/src/UserDefaults.cpp:23:
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/MiscCommon/FindCfgFile.h:11:
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/MiscCommon/SysHelper.h:33:
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/MiscCommon/MiscUtils.h:14:
In file included from /usr/local/opt/boost/include/boost/process.hpp:24:
In file included from /usr/local/opt/boost/include/boost/process/async_system.hpp:22:
In file included from /usr/local/opt/boost/include/boost/process/child.hpp:21:
In file included from /usr/local/opt/boost/include/boost/process/detail/child_decl.hpp:30:
/usr/local/opt/boost/include/boost/process/detail/posix/wait_for_exit.hpp:61:7: error: expected unqualified-id
    ::sigaddset(&sigset, SIGCHLD);
      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/signal.h:122:31: note: expanded from macro 'sigaddset'
#define sigaddset(set, signo)   (*(set) |= __sigbits(signo), 0)
                                ^
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/dds-user-defaults/src/UserDefaults.cpp:23:
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/MiscCommon/FindCfgFile.h:11:
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/MiscCommon/SysHelper.h:33:
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/MiscCommon/MiscUtils.h:14:
In file included from /usr/local/opt/boost/include/boost/process.hpp:24:
In file included from /usr/local/opt/boost/include/boost/process/async_system.hpp:22:
In file included from /usr/local/opt/boost/include/boost/process/child.hpp:21:
In file included from /usr/local/opt/boost/include/boost/process/detail/child_decl.hpp:30:
/usr/local/opt/boost/include/boost/process/detail/posix/wait_for_exit.hpp:87:26: error: no member named 'sigtimedwait' in the global namespace
        auto ret_sig = ::sigtimedwait(&sigset, nullptr, &ts);
                       ~~^
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/dds-user-defaults/src/UserDefaults.cpp:23:
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/MiscCommon/FindCfgFile.h:11:
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/MiscCommon/SysHelper.h:33:
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/MiscCommon/MiscUtils.h:14:
In file included from /usr/local/opt/boost/include/boost/process.hpp:25:
In file included from /usr/local/opt/boost/include/boost/process/group.hpp:32:
/usr/local/opt/boost/include/boost/process/detail/posix/wait_group.hpp:64:7: error: expected unqualified-id
    ::sigemptyset(&sigset);
      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/signal.h:125:26: note: expanded from macro 'sigemptyset'
#define sigemptyset(set)        (*(set) = 0, 0)
                                ^
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/dds-user-defaults/src/UserDefaults.cpp:23:
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/MiscCommon/FindCfgFile.h:11:
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/MiscCommon/SysHelper.h:33:
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/MiscCommon/MiscUtils.h:14:
In file included from /usr/local/opt/boost/include/boost/process.hpp:25:
In file included from /usr/local/opt/boost/include/boost/process/group.hpp:32:
/usr/local/opt/boost/include/boost/process/detail/posix/wait_group.hpp:65:7: error: expected unqualified-id
    ::sigaddset(&sigset, SIGCHLD);
      ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr/include/signal.h:122:31: note: expanded from macro 'sigaddset'
#define sigaddset(set, signo)   (*(set) |= __sigbits(signo), 0)
                                ^
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/dds-user-defaults/src/UserDefaults.cpp:23:
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/MiscCommon/FindCfgFile.h:11:
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/MiscCommon/SysHelper.h:33:
In file included from /Users/awegrzyn/alice/sw/SOURCES/DDS/master/0/MiscCommon/MiscUtils.h:14:
In file included from /usr/local/opt/boost/include/boost/process.hpp:25:
In file included from /usr/local/opt/boost/include/boost/process/group.hpp:32:
/usr/local/opt/boost/include/boost/process/detail/posix/wait_group.hpp:90:17: error: no member named 'sigtimedwait' in the global namespace
        ret = ::sigtimedwait(&sigset, nullptr, &ts);
              ~~^
6 errors generated.
```

**Comment by pzhristov** [comment]
This is a general problem in boost 1.69 on MacOS 10.14. I am attaching a patch that worked for me.
[patch.txt](https://github.com/alisw/alidist/files/3156226/patch.txt)

**Comment by AnarManafov** [comment]
@awegrzyn , it's not DDS, rather BOOST 1.69 is incomparable with MacOS. https://stackoverflow.com/questions/54911422/are-some-libraries-in-boost-1-69-not-compatible-with-macos

**Comment by awegrzyn** [comment]
@AnarManafov Thanks for the explanation

**Comment by mpuccio** [comment]
Hi, so what is the recommended way to compile boost on mac? Using this PR, the patch that Peter kindly provided or something else?

**Comment by Barthelemy** [comment]
if this PR were accepted you would be able to build boost 1.68 on your mac : https://github.com/alisw/alidist/pull/1607

**Comment by ktf** [comment]
Given 1.70.0 is out, how about we move directly to it?

**Comment by Barthelemy** [comment]
yes, please, and DDS at the same time ?

**Comment by Barthelemy** [comment]
any news ?

**Comment by ktf** [comment]
I hijacked @dberzano PR and moved it to use 1.70.0, as discussed. This will not be sufficient (we need to bump a bunch of other stuff) but at least will get us started.

**Comment by ktf** [comment]
I think we will need some perturbation over #1715 for this.

**Comment by dennisklein** [comment]
I tried today to get your stack compiled with Boost 1.70, this is how far I got:

I used the following `master` from alidist:
```
╭─dklein@gamma ~/alice/alidist ‹ruby-2.6.0› ‹master*›
╰─➤  git --no-pager show HEAD
commit 1ebac65a39c465e83faee5fbe4f6df3a5afa6c25 (HEAD -> master, origin/master, origin/HEAD)
Author: Kostas Alexopoulos <kostorr.alex@gmail.com>
Date:   Mon Jul 1 12:49:43 2019 +0200

    Bump ReadoutCard to v0.11.1 (#1710)

diff --git a/readoutcard.sh b/readoutcard.sh
index 6c33733..cf76f34 100644
--- a/readoutcard.sh
+++ b/readoutcard.sh
@@ -1,6 +1,6 @@
 package: ReadoutCard
 version: "%(tag_basename)s"
-tag: v0.11.0
+tag: v0.11.1
 requires:
   - boost
   - "GCC-Toolchain:(?!osx)"
```

The changes I needed to do so far:

```diff
diff --git a/arrow.sh b/arrow.sh
index 27c336b..9bdf38e 100644
--- a/arrow.sh
+++ b/arrow.sh
@@ -55,6 +55,7 @@ cmake $SOURCEDIR/cpp                                         \
       -DARROW_WITH_ZLIB=ON                                   \
       -DARROW_NO_DEPRECATED_API=ON                           \
       ${BOOST_ROOT:+-DBOOST_ROOT=${BOOST_ROOT}}              \
+      -DBoost_NO_BOOST_CMAKE=ON                              \
       -DCMAKE_INSTALL_PREFIX=$INSTALLROOT                    \
       -DARROW_PYTHON=OFF

diff --git a/asiofi.sh b/asiofi.sh
index 224d7ab..48711d3 100644
--- a/asiofi.sh
+++ b/asiofi.sh
@@ -1,6 +1,6 @@
 package: asiofi
 version: "%(tag_basename)s"
-tag: v0.3.1
+tag: v0.4.2
 source: https://github.com/FairRootGroup/asiofi
 requires:
   - boost
diff --git a/boost.sh b/boost.sh
index 322edf5..8913937 100644
--- a/boost.sh
+++ b/boost.sh
@@ -1,7 +1,7 @@
 package: boost
 version: "%(tag_basename)s"
-tag: v1.68.0
-source: https://github.com/alisw/boost.git
+tag: v1.70.0
+source: https://github.com/dennisklein/boost.git
 requires:
   - "GCC-Toolchain:(?!osx)"
   - "Python-modules:(?!osx)"
diff --git a/common-o2.sh b/common-o2.sh
index 83b5327..6e320f6 100644
--- a/common-o2.sh
+++ b/common-o2.sh
@@ -22,6 +22,7 @@ esac
 cmake $SOURCEDIR                               \
       -DCMAKE_INSTALL_PREFIX=$INSTALLROOT      \
       ${BOOST_ROOT:+-DBOOST_ROOT=$BOOST_ROOT}  \
+      -DBUILD_SHARED_LIBS=ON \
       -DCMAKE_EXPORT_COMPILE_COMMANDS=ON

 cp ${BUILDDIR}/compile_commands.json ${INSTALLROOT}
diff --git a/configuration.sh b/configuration.sh
index 8be1ed7..79ecde8 100644
--- a/configuration.sh
+++ b/configuration.sh
@@ -23,6 +23,7 @@ esac

 cmake $SOURCEDIR                                             \
       -DCMAKE_INSTALL_PREFIX=$INSTALLROOT                    \
+      -DBoost_NO_BOOST_CMAKE=ON                              \
       ${BOOST_ROOT:+-DBOOST_ROOT=$BOOST_ROOT}                \
       -Dppconsul_DIR=${PPCONSUL_ROOT}/cmake

diff --git a/dds.sh b/dds.sh
index f6ad98b..dc87e9d 100644
--- a/dds.sh
+++ b/dds.sh
@@ -1,6 +1,6 @@
 package: DDS
 version: "%(tag_basename)s"
-tag: "2.2"
+tag: "2.4"
 source: https://github.com/FairRootGroup/DDS
 requires:
   - boost
@@ -25,6 +25,8 @@ esac

 cmake $SOURCEDIR                                                         \
       -DCMAKE_INSTALL_PREFIX=$INSTALLROOT                                \
+      -DBUILD_SHARED_LIBS=ON                                             \
+      -DBoost_NO_BOOST_CMAKE=ON                                          \
       ${BOOST_ROOT:+-DBOOST_ROOT=$BOOST_ROOT -DBoost_NO_SYSTEM_PATHS=ON} \

 # Limit the number of build processes to avoid exahusting memory when building
diff --git a/fairlogger.sh b/fairlogger.sh
index 73de7e0..b27533a 100644
--- a/fairlogger.sh
+++ b/fairlogger.sh
@@ -1,6 +1,6 @@
 package: FairLogger
 version: "%(tag_basename)s"
-tag: v1.2.0
+tag: v1.5.0
 source: https://github.com/FairRootGroup/FairLogger
 build_requires:
  - CMake
diff --git a/fairmq.sh b/fairmq.sh
index e6bd2cd..1edb9d5 100644
--- a/fairmq.sh
+++ b/fairmq.sh
@@ -1,6 +1,6 @@
 package: FairMQ
 version: "%(tag_basename)s"
-tag: v1.4.2
+tag: dev
 source: https://github.com/FairRootGroup/FairMQ
 requires:
  - boost
diff --git a/fairroot.sh b/fairroot.sh
index f4351f2..5ab4f7a 100644
--- a/fairroot.sh
+++ b/fairroot.sh
@@ -1,7 +1,7 @@
 package: FairRoot
 version: "%(short_hash)s"
-tag: "c672f280ec8382cb2dc10b85c5fa4e64e13c5975"
-source: https://github.com/FairRootGroup/FairRoot
+tag: "support_boost_170"
+source: https://github.com/dennisklein/FairRoot
 requires:
   - generators
   - simulation
diff --git a/flatbuffers.sh b/flatbuffers.sh
index 5b782a5..d55ed31 100644
--- a/flatbuffers.sh
+++ b/flatbuffers.sh
@@ -1,5 +1,5 @@
 package: flatbuffers
-version: v1.10.0
+version: v1.11.0
 source: https://github.com/google/flatbuffers
 requires:
   - zlib
diff --git a/mcsteplogger.sh b/mcsteplogger.sh
index d7d0603..453b958 100644
--- a/mcsteplogger.sh
+++ b/mcsteplogger.sh
@@ -13,6 +13,7 @@ build_requires:
 #!/bin/bash -e
 cmake $SOURCEDIR -DCMAKE_INSTALL_PREFIX=$INSTALLROOT   \
           ${CXXSTD:+-DCMAKE_CXX_STANDARD=$CXXSTD}      \
+          -DBoost_NO_BOOST_CMAKE=ON                    \
           -DROOT_DIR=${ROOT_ROOT}                      \
           -DCMAKE_EXPORT_COMPILE_COMMANDS=ON

diff --git a/monitoring.sh b/monitoring.sh
index b55ca55..57168ec 100644
--- a/monitoring.sh
+++ b/monitoring.sh
@@ -24,6 +24,7 @@ fi

 cmake $SOURCEDIR                                              \
       -DCMAKE_INSTALL_PREFIX=$INSTALLROOT                     \
+      -DBoost_NO_BOOST_CMAKE=ON                               \
       ${BOOST_VERSION:+-DBOOST_ROOT=$BOOST_ROOT}                 \
       ${APMON_CPP_VERSION:+-DAPMON_ROOT=$APMON_CPP_ROOT}         \
       -DCMAKE_EXPORT_COMPILE_COMMANDS=ON
```

Comments:
* FairMQ `dev` is meanwhile available as `v1.4.5`
* For bumping Boost to 1.70 I have provided alisw/boost#2
* All packages where I needed to add `-DBoost_NO_BOOST_CMAKE=ON` need work in their CMake code to support the Boost 1.70 CMake package (or leave it disabled ;) )
* All packages where I added `-DBUILD_SHARED_LIBS=ON` are already using modern enough CMake code for the new Boost 1.70 CMake package. Unfortunately, this option is needed given the way we configure Boost. This is [already acknowledged by upstream boost as a bug](https://github.com/boostorg/boost_install/issues/5) and will be fixed in Boost 1.71. Anyways, if you decide to add statically to your CMake code, be aware of [its full meaning](https://cmake.org/cmake/help/latest/variable/BUILD_SHARED_LIBS.html) (I recommend setting only a default `if(NOT DEFINED BUILD_SHARED_LIBS)` to be on the safe side).
* DDS has already addressed the Boost 1.70 CMake issues in their `dev` branch.
* Bumping asiofi to `v0.4.2` might lead to issues on some platforms due to a configuration issue with the included `folly`. There are three ways to fix this:
  1. Provide a proper `folly` installation via `-DCMAKE_PREFIX_PATH` or `-DFOLLY_ROOT` when configuring asiofi. (New alidist recipes ...)
  2. Do not build asiofi everywhere, especially on MacOS it is really useless (and not tested).
  3. Help me by finding a good way of repairing the builtin folly in asiofi (It is not a full folly installation but a handpicked selection of headers files with just the types I depend on, works only in combination with preprocessor definition `FOLLY_NO_CONFIG` - maybe this is not the best idea and we can only really go with option 1...)
* FairRoot is not building for me yet in your stack. Some issue with pythia6 library naming. Investigating..

All packages from the FairRootGroup org - FairRoot (once FairRootGroup/FairRoot#898 is merged), FairMQ `v1.4.5`, asiofi `v0.4.2`, and FairLogger `v1.5.0` - deal with Boost 1.70 CMake issues in the following way:
* By default `Boost_NO_BOOST_CMAKE` is set to `ON` if `CMAKE_VERSION` is less then `3.15` (You can override via `-D` on the command line)
* All those repos should support Boost 1.70 config if you decide to override the default. I hope I did not overlook a case where they do not, if you find anything not working, please file an issue with us. Thanks!

**Comment by ktf** [comment]
@dennisklein thanks, however I have hard time making your comment actionable. Could you open a PR for the things you think can already go in our alidist and that are required by the boost 1.70.0 migration?

**Comment by dennisklein** [comment]
> @dennisklein thanks, however I have hard time making your comment actionable. Could you open a PR for the things you think can already go in our alidist and that are required by the boost 1.70.0 migration?

Thing is, what is required all depends:
1. Do you insist on using the Boost `1.70` CMake package?
2. Do you insist on pushing this before `CMake 3.15` comes out (it is already `rc3`)?

**Comment by ktf** [comment]
Closing this in favour of #1776 which seems to be almost there.

**Pull request closed without merging**